tensorboard/plugins/image/summary.py

```diff
@@ -0,0 +1,135 @@
+                                tf.assert_type(images, tf.uint8),
+                                tf.assert_non_negative(max_outputs)]):
+    limited_images = images[:max_outputs]
+    encoded_images = tf.map_fn(tf.image.encode_png, limited_images,
+                               dtype=tf.string,
+                               name='encode_each_image')
+    image_shape = tf.shape(images)
+    dimensions = tf.stack([tf.as_string(image_shape[1], name='width'),
+                           tf.as_string(image_shape[2], name='height')],
+                          name='dimensions')
+    tensor = tf.concat([dimensions, encoded_images], axis=0)
+    return tf.summary.tensor_summary(name='image_summary',
+                                     tensor=tensor,
+                                     collections=collections,
+                                     summary_metadata=summary_metadata)
+
+
+def pb(name, images, max_outputs=3, display_name=None, description=None):
+  """Create an image summary protobuf.
+
+  This behaves as if you were to create an `op` with the same arguments
+  (wrapped with constant tensors where appropriate) and then execute
+  that summary op in a TensorFlow session.
+
+  Arguments:
+    name: A unique name for the generated summary, including any desired
+      name scopes.
+    images: An `np.array` representing pixel data with shape
+      `[k, w, h, c]`, where `k` is the number of images, `w` and `h` are
+      the width and height of the images, and `c` is the number of
+      channels, which should be 1, 3, or 4.
+    max_outputs: Optional `int`. At most this many images will be
+      emitted. If more than this many images are provided, the first
+      `max_outputs` many images will be used and the rest silently
+      discarded.
+    display_name: Optional name for this summary in TensorBoard, as a
+      `str`. Defaults to `name`.
+    description: Optional long-form description for this summary, as a
+      `str`. Markdown is supported. Defaults to empty.
+
+  Returns:
+    A `tf.Summary` protobuf object.
+  """
+  images = np.array(images).astype(np.uint8)
+  if images.ndim != 4:
+    raise ValueError('Shape %r must have rank 4' % (images.shape, ))
+
+  limited_images = images[:max_outputs]
+  encoded_images = [_encode_png(image) for image in limited_images]
+  (width, height) = (images.shape[1], images.shape[2])
+  content = [str(width), str(height)] + encoded_images
+  tensor = tf.make_tensor_proto(content, dtype=tf.string)
+
+  if display_name is None:
+    display_name = name
+  summary_metadata = metadata.create_summary_metadata(
+      display_name=display_name, description=description)
+
+  summary = tf.Summary()
+  summary.value.add(tag='%s/image_summary' % name,
+                    metadata=summary_metadata,
+                    tensor=tensor)
+  return summary
+
+
+def _encode_png(image):
+  # TODO(@wchargin): Pick a third-party PNG backend and implement this.
```

Comment:
We could always do this later on:

```py
try:
  import pillow
  encode_png = pillow.png_encoder_function
except ImportError:
  import tensorboard.png_encoder
  encode_png = png_encoder.encode_png
```

But we shouldn't need to. My understanding of things is that people don't want to use TensorFlow in their own code, i.e. we shouldn't make it a mandatory part of TensorBoard's API contracts. If TensorFlow is a transitive dependency, they don't have to think about it; it just happens in the background. Since I added a cherry-pick to make `import tensorflow` in 1.3.x go 2x faster (and I intend to make 1.4.x go 10x faster) I don't see why anyone would object to that transitive dependency for reasons other than philosophical.

The reason why we're probably going to need it on the serving side is because I'm considering having TensorBase not store PNG encoded images, and instead store them as tensors, and then have the WSGI handler encode them to PNG on the fly. Also Beholder encodes PNGs while serving.